crates/polars-io/src/partition.rs

```diff
@@ -139,100 +139,152 @@ pub fn write_partitioned_dataset<S>(
+            .add(b'/')
+            .add(b'=')
+            .add(b':')
+            .add(b' ');
 
-        fn get_path_for_index(i: usize) -> String {
-            // Use a fixed-width file name so that it sorts properly.
-            format!("{:013x}.parquet", i)
+        move |df: &DataFrame| {
+            let cols = df.get_columns();
+
+            partition_by_col_idx
+                .iter()
+                .map(|&i| {
+                    let s = &cols[i].slice(0, 1).cast(&DataType::String).unwrap();
+
+                    format!(
+                        "{}={}",
+                        s.name(),
+                        percent_encoding::percent_encode(
+                            s.str()
+                                .unwrap()
+                                .get(0)
+                                .unwrap_or("__HIVE_DEFAULT_PARTITION__")
+                                .as_bytes(),
+                            CHAR_SET
+                        )
+                    )
+                })
+                .collect::<Vec<_>>()
+                .join("/")
         }
+    };
 
-        for (i, slice_start) in (0..part_df.height()).step_by(rows_per_file).enumerate() {
-            let f = std::fs::File::create(dir.join(get_path_for_index(i)))?;
+    let base_path = path;
+    let groups = df.group_by(partition_by)?.take_groups();
 
-            file_write_options
-                .to_writer(f)
-                .finish(&mut part_df.slice(slice_start as i64, rows_per_file))?;
-        }
-    }
+    let init_part_base_dir = |part_df: &DataFrame| {
+        let path_part = get_hive_path_part(part_df);
+        let dir = base_path.join(path_part);
+        std::fs::create_dir_all(&dir)?;
 
-    Ok(())
-}
+        PolarsResult::Ok(dir)
+    };
 
-/// Creates an iterator of (hive partition path, DataFrame) pairs, e.g.:
-/// ("a=1/b=1", DataFrame)
-fn get_hive_partitions_iter<'a, S>(
-    df: &'a DataFrame,
-    partition_by: &'a [S],
-) -> PolarsResult<Box<dyn Iterator<Item = (String, DataFrame)> + 'a>>
-where
-    S: AsRef<str>,
-{
-    let schema = df.schema();
-
-    let partition_by_col_idx = partition_by
-        .iter()
-        .map(|x| {
-            let Some(i) = schema.index_of(x.as_ref()) else {
-                polars_bail!(ColumnNotFound: "{}", x.as_ref())
-            };
-            Ok(i)
-        })
-        .collect::<PolarsResult<Vec<_>>>()?;
-
-    let get_hive_path_part = move |df: &DataFrame| {
-        const CHAR_SET: &percent_encoding::AsciiSet = &percent_encoding::CONTROLS
-            .add(b'/')
-            .add(b'=')
-            .add(b':')
-            .add(b' ');
+    fn get_path_for_index(i: usize) -> String {
+        // Use a fixed-width file name so that it sorts properly.
+        format!("{:08x}.parquet", i)
+    }
 
-        let cols = df.get_columns();
+    let get_n_files_and_rows_per_file = |part_df: &DataFrame| {
+        let n_files = (part_df.estimated_size() / chunk_size).clamp(1, 0xffff_ffff);
```

Comment:
Reduced the max number of files per-directory from `4_503_599_627_370_495` to `4_294_967_295`, don't think we needed that many 😅. This also reduces the length of the file name by 5 characters, which should help with sorting.

Reply:
Nah, I doubt we'll hit that in our lifetime. :')